ext/ddtrace_profiling_native_extension/heap_recorder.c

```diff
@@ -0,0 +1,489 @@
+      .obj = obj,
+      .weight = weight,
+      .free = false,
+      .skip = false,
+  });
+}
+
+static void enqueue_free(heap_recorder *heap_recorder, VALUE obj) {
+  enqueue_sample(heap_recorder, (sample) {
+      .stack = NULL,
+      .obj = obj,
+      .weight = 0,
+      .free = true,
+      .skip = false,
+  });
+}
+
+void start_heap_allocation_recording(heap_recorder* heap_recorder, VALUE new_obj, unsigned int weight, ddog_CharSlice *class_name) {
+  fprintf(stderr, "Started recording allocation of %lu with weight %u\n", new_obj, weight);
+  partial_heap_recording *active_recording = &heap_recorder->active_recording;
+  active_recording->obj = new_obj;
+  active_recording->weight = weight;
+  active_recording->class_name = class_name;
+}
+
+void end_heap_allocation_recording(struct heap_recorder *heap_recorder, ddog_prof_Slice_Location locations) {
+  // TODO: Make use of active_recording->class_name
+  partial_heap_recording *active_recording = &heap_recorder->active_recording;
+
+  VALUE new_obj = active_recording->obj;
+  if (!new_obj) {
+    // Recording ended without having been started?
+    rb_raise(rb_eRuntimeError, "Ended a heap recording that was not started");
+    return;
+  }
+
+  int weight = active_recording->weight;
+
+  // From now on, mark active recording as invalid so we can short-circuit at any point and
+  // not end up with a still active recording. new_obj still holds the object for this recording
+  active_recording->obj = 0;
+
+  heap_stack *heap_stack = heap_stack_init(locations);
+  int error = pthread_mutex_trylock(&heap_recorder->records_mutex);
+  if (!error) {
+    // We were able to get a lock to heap_records so lets flush any pending samples
+    // that might have been queued previously before adding this new one.
+    flush_queue(heap_recorder);
+  } else {
+    // We weren't able to get a lock, so enqueue this sample for later processing
+    // and end early
+    if (error == EBUSY) {
+      enqueue_allocation(heap_recorder, heap_stack, new_obj, weight);
+    } else {
+      ENFORCE_SUCCESS_GVL(error)
+    }
+    return;
+  }
+
+  // If we got this far, we got a write lock so we can commit the record
+  commit_allocation(heap_recorder, heap_stack, new_obj, weight);
+
+  pthread_mutex_unlock(&heap_recorder->records_mutex);
+}
+
+void record_heap_free(heap_recorder *heap_recorder, VALUE obj) {
+  object_record *object_record = NULL;
+  st_lookup(heap_recorder->object_records, (st_data_t) obj, (st_data_t*) &object_record);
+
+  if (object_record == NULL) {
+    // we don't seem to be tracking this object on the table atm
+    // check if the allocation sample is in the queue
+    for (size_t i = 0; i < heap_recorder->queued_samples_len; i++) {
+      sample *queued_sample = &heap_recorder->queued_samples[i];
+      if (queued_sample->obj == obj) {
+        queued_sample->skip = true;
+        break;
+      }
+    }
+
+    // free of an untracked object, return early
+    return;
+  }
```

Comment:
I guess rare situation, but if for some reason the background thread really gets behind, it seems possible that object at location A is allocated, then freed, then location A is reused while the queue remains unflushed.

In such a case, this check be incorrect, since we'd skip the first allocation, but not the second one.